File: quickstart_guides/arrays_strings/python/wildcard_matching.py
"""
Title: Wildcard matching

Problem: 
    Given an input string (s) and a pattern (p), implement wildcard pattern
    matching with support for '?' and '*'.

    '?' Matches any single character.  '*' Matches any sequence of characters
    (including the empty sequence). The matching should cover the entire input
    string (not partial).

    Note:

    s could be empty and contains only lowercase letters a-z.  p could be empty
    and contains only lowercase letters a-z, and characters like ? or *.

Execution: python wildcard_matching.py
"""
import unittest


def wildcard_matching(a: str, b: str) -> bool:
    if len(a) == len(b) == 0:
        return True
    res = False
    seen = set()
    dfs(a, b, 0, 0, seen, res)
    return res


def dfs(s, p, i, j, seen, res) -> None:
    if (i, j) in seen:
        return

    # Recursion exit condition.
    if j > len(p) - 1:
        return

    # Generate multiple cases.
    if p[j] == "*":
        for i in range(i, len(s)):
            if (i, j + 1) not in seen:
                dfs(s, p, i, j + 1, seen, res)
                seen.add((i, j + 1))
            else:
                return

    elif i < len(s) and (p[j] == "?" or s[i] == p[j]):
        if (i + 1, j + 1) not in seen:
            dfs(s, p, i + 1, j + 1, seen, res)
            seen.add((i + 1, j + 1))
        else:
            return


class TestWildcardMatching(unittest.TestCase):
    """Unit test for wildcard_matching."""

    def test_1(self):
        a = "aa"
        p = "a"
        self.assertEqual(wildcard_matching(a, p), False)

    def test_2(self):
        a = "acdcb"
        p = "a*c?b"
        self.assertEqual(wildcard_matching(a, p), False)


if __name__ == "__main__":
    unittest.main()

File: quickstart_guides/arrays_strings/java/WildcardMatching.java
/*
 *   Title: Wildcard Matching
 *
 *   Problem:
 *       Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*'.
 *
 *       '?' Matches any single character.
 *       '*' Matches any sequence of characters (including the empty sequence).
 *       The matching should cover the entire input string (not partial).
 *
 *       Note:
 *
 *       s could be empty and contains only lowercase letters a-z.
 *       p could be empty and contains only lowercase letters a-z, and characters like ? or *.
 *
 *   Execution: javac WildcardMatching.java && java WildcardMatching
 */

import java.util.*;

public class WildcardMatching {
    public static boolean wildcardMatching(String s, String p) {
        boolean[][] dp = new boolean[s.length() + 1][p.length() + 1];
        dp[0][0] = true;
        for (int i = 0; i < p.length(); i++) {
            if(p.charAt(i) == '*' && dp[0][i]){
                dp[0][i+1] = true;
            }
        }
        for (int i = 0; i < s.length(); i++) {
            for (int j = 0; j < p.length(); j++) {
                if(s.charAt(i) == p.charAt(j) || p.charAt(j) == '?'){
                    dp[i+1][j+1] = dp[i][j];
                }
                if(p.charAt(j) == '*'){
                    dp[i + 1][j + 1] = dp[i][j + 1] || dp[i + 1][j];
                }
            }
        }
        return dp[s.length()][p.length()];
    }
    
    public static void main(String[] args) {
        assert wildcardMatching("aa", "a") == false;
        assert wildcardMatching("aa", "*") == true;
        assert wildcardMatching("cb", "?a") == false;
        assert wildcardMatching("adceb", "*a*b") == true;
        assert wildcardMatching("acdcb", "a*c?b") == false;

        System.out.println("Passed all test cases");
    }
}
